src/upnp.cpp

```diff
@@ -277,7 +277,7 @@ void upnp::discover_device_impl()
 
 // returns a reference to a mapping or -1 on failure
 port_mapping_t upnp::add_mapping(portmap_protocol const p, int const external_port
-	, tcp::endpoint const local_ep)
+	, tcp::endpoint const& local_ep)
```

Comment:
this change makes sense.
Every other change in this commit (after a cursory scan) look wrong. I'm pretty sure I *want* to pass any type that can be efficiently moved, and that the function want a copy of, by value. this would include smart pointers and vectors for instance.